### ISSUE TITLE
Fix postsubmit reruns triggered from GitHub UI

### ISSUE THREAD

**Issue by nehalvpatel** [flutter/flutter#116739]
Addresses https://github.com/flutter/flutter/issues/116226

The current plan is to use the [checkrun listing API](https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#list-check-runs-for-a-git-reference) call to find the existing checkrun:
```bash
gh api -H "Accept: application/vnd.github+json" "/repos/flutter/plugins/commits/614d7c0df1eb03d2579118b4a0c3032dbedfa364/check-runs?check=Mac_x64%20ios_platform_tests_4_of_4%20master"
```

**Comment by godofredoc** [comment]
This is impacting plugins, packages and dart-internal

**Comment by godofredoc** [comment]
Related to this bug:

* https://github.com/flutter/flutter/issues/115730

**Comment by nehalvpatel** [comment]
Update rescheduleUsingCheckRunEvent with logic that differentiates presubmit and postsubmit
- `getTryBuilds` + `_createPresubmitScheduleBuild`
- `getProdBuilds` + `_createPostsubmitScheduleBuild`

**Comment by nehalvpatel** [comment]
The tricky part is how to know if it's a postsubmit or presubmit checkrun.

Tentative plan:

One way is to add a datastore check here when the `rerequested` event is received:
- https://github.com/flutter/cocoon/blob/fa3d7c16d080f6f067ce9513f38a52c59990454c/app_dart/lib/src/service/scheduler.dart#L440

It would be relying on the fact that right now, only postsubmit tasks are stored in the datastore. If found in the datastore, it is postsubmit, and if not, it is presubmit.

But it could be implemented as this temporarily, until I refactor cocoon models/services API next quarter.
- This unblocks the plugins team: https://github.com/flutter/flutter/issues/116226#issuecomment-1336009821

First, I would land a small refactor PR that breaks out common functionality for `rescheduleUsingCheckRunEvent` into separate methods for presubmit and postsubmit:
- https://github.com/flutter/cocoon/blob/a947ac7f2d5fd66020c1ea9a230af1c459db2ba6/app_dart/lib/src/service/luci_build_service.dart#L365

Then, I would land another PR that adds the datastore check method, updates the `rerequested` event, and adds tests.

**Comment by godofredoc** [comment]
There are a few complications with the first approach. Apparently there is no way to identify a database entry with the information you are receiving in the event generated if the retry is generated from GitHub.

I believe the easier temporary fix is to focus on retries created from the flutter-dashboard and only fix the update of the checkrun.

Are there any api curl commands you have been using to validate the workflows? we'll need something like:

* On a landed plugins commit, send a curl request to create a new check run with one of the existing names.
* Send another request to mark it as red.
* Create a new check run with the same name and now mark it as green

If the manual workflow works we only need to fix the checkrun update logic.

**Comment by nehalvpatel** [comment]
### Manual workflow:

<details>
<summary>I was able to run the manual workflow with curl on my test repo and it worked. I've attached screenshots here.</summary>

## 1 - curl - Create new check run
<img width="953" alt="Screenshot 2022-12-16 at 12 52 22 PM - curl - check run started" src="https://user-images.githubusercontent.com/159021/208189116-54a93915-abe7-4f06-b037-dd3a21502d18.png">

## 2 - GitHub UI shows check run in progress
<img width="1021" alt="Screenshot 2022-12-16 at 12 52 41 PM - curl check run in progress on github" src="https://user-images.githubusercontent.com/159021/208189125-af6f35e3-5649-4fd3-90d9-78c7acacf872.png">

## 3 - curl - Mark check run as failed
<img width="950" alt="Screenshot 2022-12-16 at 12 55 21 PM - curl check run failure" src="https://user-images.githubusercontent.com/159021/208189130-97bf6bb5-3c95-4690-9773-a849e0a5acb8.png">

## 4 - GitHub UI shows check run as failed
<img width="1025" alt="Screenshot 2022-12-16 at 12 55 35 PM - curl check run fail on github" src="https://user-images.githubusercontent.com/159021/208189134-74a9c9d2-f821-4449-b9c0-7fddce03edc8.png">

## 5 - curl - Create new check run with the same name
<img width="947" alt="Screenshot 2022-12-16 at 12 57 08 PM - curl create new one with same name" src="https://user-images.githubusercontent.com/159021/208189137-8bf9aa1e-939a-41d2-bc46-6474d9cc493c.png">

## 6 - GitHub UI shows check run in progress
<img width="1024" alt="Screenshot 2022-12-16 at 12 57 25 PM - curl check run in progress again on github" src="https://user-images.githubusercontent.com/159021/208189141-c107623b-d93f-41d5-b9cd-e9600a7a59e8.png">

## 7 - curl - Mark check run as successful
<img width="950" alt="Screenshot 2022-12-16 at 12 58 22 PM - curl check run success" src="https://user-images.githubusercontent.com/159021/208189142-7f7f85fa-0d13-4c9e-ab7e-aaeb6d64d856.png">

## 8 - GitHub UI shows check run as successful
<img width="1023" alt="Screenshot 2022-12-16 at 12 58 40 PM - curl check run success on github" src="https://user-images.githubusercontent.com/159021/208189144-d6b89cf4-39dd-462e-81e4-990d64ad7a4e.png">
</details>

### flutter-dashboard retries might not be the issue:

I checked the webhook logs for the commit mentioned in https://github.com/flutter/flutter/issues/116226, and it looks like someone had attempted a retry from the GitHub UI before the attempt from flutter-dashboard. And since the GitHub postsubmit reruns are not wired up right now, it keeps failing and re-creating new checkruns. Once a task/check is in this state, it cannot recover even if you do a retry from flutter-dashboard.
- One clue is the checkrun linked in the issue is no longer available: https://github.com/flutter/plugins/runs/9782441143
- I also observed this loop by re-running other failed postsubmit check runs from GitHub.
- Some more info is in the screenshots below.

### Timeline:
- **Nov 28, 2022, 4:59 PM PST** - https://github.com/flutter/plugins/commit/614d7c0df1eb03d2579118b4a0c3032dbedfa364 was committed and some time later, the `Mac_x64 ios_platform_tests_4_of_4 master` task failed.

- **Nov 29, 2022, 7:16 AM PST** - The failed check run was re-run from the GitHub UI. The task is now in a glitched state where the infra keeps trying to schedule it, fails, and repeats.

  - <img width="1448" alt="Screenshot 2022-12-16 at 1 39 13 PM" src="https://user-images.githubusercontent.com/159021/208193701-33a952eb-946e-46ee-b65b-a811528641b0.png">

  - ![BuABvRtjRyvinb4](https://user-images.githubusercontent.com/159021/208196240-78ca4112-8923-49e2-b1fa-1180acb30a8b.png)

- **Nov 29, 2022, 12:29 PM PST** - The failed check run was re-run from flutter-dashboard. This is assumed to be the culprit, but is actually the only reason a 2nd attempt was launched.

  - <img width="1429" alt="Screenshot 2022-12-16 at 1 44 33 PM" src="https://user-images.githubusercontent.com/159021/208194165-19dece69-ff81-4a53-8b89-0dfbc2d65aa1.png">
  - ![3W8oxrLmSSmK6eS](https://user-images.githubusercontent.com/159021/208195610-303275ab-0a78-40cc-96de-6d875cf33e41.png)

- **Nov 29, 2022, 12:52 PM PST** - The 2nd attempt was successful, but its result was overridden by the still-looping checkrun recreations from the GitHub attempt.

  - ![7uoirR8RfzvVaPj](https://user-images.githubusercontent.com/159021/208196048-44e9cef8-376b-43a7-bbfb-ddc7e0e51956.png)

  - ![5Yicaowjss5tKVi](https://user-images.githubusercontent.com/159021/208196137-f501e61d-b32a-4acd-bc8e-748b53622bd2.png)

**Comment by godofredoc** [comment]
To make things ever more complicated, if a post-submit task fails it will be automatically retried.

Why is important to know if the run is for pre/post submit?

**Comment by nehalvpatel** [comment]
I believe the main reason is there are two discrete code paths to take to launch the new attempt/build, depending on if the run is pre/post submit.
- If presubmit: we would use `getTryBuilds` and `_createPresubmitScheduleBuild`, as currently seen within `rescheduleUsingCheckRunEvent`
- If postsubmit: we would use `getProdBuilds` and `_createPostsubmitScheduleBuild`

To simplify things, I'd probably break out some common functionality into some utility functions.

**Comment by godofredoc** [comment]
Let's check if we can simplify even more those methods. Do we get a PR in the same slug on event updates for presubmit?

**Comment by nehalvpatel** [comment]
I checked a bunch of logged `rerequested` events for different repos and here's what I found:

**presubmit**:
- pull_requests is empty (flutter/flutter)
- pull_requests is empty (flutter/engine)
- pull_requests is empty (flutter/plugins)
- pull_requests is empty (flutter/packages)

**postsubmit**:
- pull_requests has random repositories (flutter/plugins)
  - these are people who made a PR in their fork with the source being flutter/flutter:master
- couldn't find an example for (flutter/packages)

**Comment by nehalvpatel** [comment]
One alternative I could think of is to pre-empt the upcoming refactor and add a new column to the datastore that identifies the task as presubmit/postsubmit/etc. Right now, every task would just have `postsubmit` as the value in that column. The logic for this PR would check that column instead of just the existence of a task, as considered above. So that in the future, when `presubmit` tasks are also in the datastore, we wouldn't need to remember to come back to fix the hack.

Another option is to warn people not to perform postsubmit re-runs from the GitHub UI. This would be a temporary solution until this is refactored next quarter.

**Comment by nehalvpatel** [comment]
Still tracking this. I'll be discussing this further with the team this week and post updates.

**Comment by godofredoc** [comment]
I think I have an idea to identify if it is presubmit/postsubmit that avoids having to save state.

We can use the GoB API to get the commit info passed from the event, if info is returned from GoB then it is a post-submit otherwise is a presubmit.

**Comment by godofredoc** [comment]
It will need to extend https://cs.opensource.google/flutter/cocoon/+/main:app_dart/lib/src/service/gerrit_service.dart to support https://gerrit-review.googlesource.com/Documentation/rest-api-projects.html#get-commit.

**Comment by nehalvpatel** [comment]
That seems promising! I'll take a look.

**Comment by nehalvpatel** [comment]
Handed off oncall this morning and I'm continuing on this PR this week.

**Comment by nehalvpatel** [comment]
PR is in review now, awaiting changes from my end.

https://github.com/flutter/cocoon/pull/2387

**Comment by nehalvpatel** [comment]
The fix was merged and deployed! https://github.com/flutter/cocoon/pull/2387#issuecomment-1404259901
